feat(openapi): enhance support for tuple return type validation

philiptzou opened the pull request:
<!-- markdownlint-disable MD041 MD043 -->
**Issue number:** #5034

## Summary

### Changes

This PR fixes the issue that OpenAPI validation failed to validate two-element tuple returns, which is allowed by the `APIGateway`.

### User experience

Below test fails before this PR but passes after this PR:

```python3
from pydantic import BaseModel
from aws_lambda_powertools.event_handler import ALBResolver

app = ALBResolver(enable_validation=True)


class Result(BaseModel):
    payload: str


@app.post('/foobar')
def foobar() -> tuple[Result, int]:
    return Result(payload='foobar'), 201


def test_foobar():
    response = app.resolve({
        'path': '/foobar',
        'httpMethod': 'POST',
        'queryStringParameters': {},
        'headers': {}
    }, None)
    assert response['statusCode'] == 201
```
## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda/python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [ ] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/.github/semantic.yml)

<details>
<summary><s>Is this a breaking change?</s></summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

philiptzou committed:
feat(openapi): enhance support for tuple return type validation

leandrodamascena commented on tests/functional/event_handler/_pydantic/test_openapi_params.py:
```diff
@@ -172,6 +172,42 @@ def handler() -> Response[Annotated[str, Body(title="Response title")]]:
+    app = APIGatewayRestResolver()
+
+    @app.get("/")
+    def handler() -> tuple[str, int]:
```
Can you pls import `Tuple` from typing? I need to migrate our tests to adopt the new type annotation, but we have some issues for now.

philiptzou replied:
Done.

leandrodamascena commented on tests/functional/event_handler/_pydantic/test_openapi_params.py:
```diff
@@ -172,6 +172,42 @@ def handler() -> Response[Annotated[str, Body(title="Response title")]]:
+    app = APIGatewayRestResolver()
+
+    @app.get("/")
+    def handler() -> tuple[Annotated[str, Body(title="Response title")], int]:
```
Can you pls import `Tuple` from typing? I need to migrate our tests to adopt the new type annotation, but we have some issues for now.

philiptzou replied:
Done.

leandrodamascena reviewed (changes requested):
Hey @philiptzou! Thanks a lot for submitting this PR. 🚀 

I see we have some problems with type checking that is breaking the CI and we need to address them before I make a final review.

leandrodamascena committed:
Merge branch 'develop' into openapi-with-tuple-returns

leandrodamascena committed:
Merge branch 'develop' into openapi-with-tuple-returns

leandrodamascena reviewed (approved):
Hey @philiptzou! Thanks for working on this PR and addressing the feedback!

I have a feeling that this PR can breaks the validation for people using dict as return. But I think this is really a bug that we need to fix.

e.g
```python
@app.get('/foobar')
def foobar() -> dict:
    return Result(payload='foobar'), 201
```

APPROVED!

philiptzou commented:
> Hey @philiptzou! Thanks for working on this PR and addressing the feedback!
> 
> I have a feeling that this PR can breaks the validation for people using dict as return. But I think this is really a bug that we need to fix.
> 
> e.g
> 
> ```python
> @app.get('/foobar')
> def foobar() -> dict:
>     return Result(payload='foobar'), 201
> ```
> 
> APPROVED!

Well, I don't think this PR will break dict as return since it still falls in the `else` branch. This example still work, but it irritates mypy anyway.
